make_prg/utils/io_utils.py

```diff
@@ -0,0 +1,219 @@
+    msa_file = str(msa_file)
+    if msa_file.endswith(".gz"):
+        handle = gzip.open(msa_file, "rt")
+        alignment = AlignIO.read(handle, alignment_format)
+        handle.close()
+    else:
+        alignment = AlignIO.read(msa_file, alignment_format)
+    for record in alignment:
+        record.seq = record.seq.upper()
+    return alignment
+
+
+def concatenate_text_files(input_filepaths, output_filepath):
+    output_filepath_parent_dir = Path(output_filepath).parent
+    os.makedirs(output_filepath_parent_dir, exist_ok=True)
+
+    with open(output_filepath, "w") as fout, fileinput.input(input_filepaths) as fin:
+        for line in fin:
+            fout.write(line)
+
+
+# Note: not unit tested
+# From https://gist.github.com/jacobtomlinson/9031697
+def remove_empty_folders(path: str, remove_root: bool = True):
+    if not os.path.isdir(path):
+        return
+
+    # remove empty subfolders
+    files = os.listdir(path)
+    for f in files:
+        fullpath = os.path.join(path, f)
+        if os.path.isdir(fullpath):
+            remove_empty_folders(fullpath)
+
+    # if folder empty, delete it
+    files = os.listdir(path)
+    if len(files) == 0 and remove_root:
+        os.rmdir(path)
+
+
+# Note: not unit tested
+def output_files_already_exist(output_prefix: str):
+    return (
+        Path(output_prefix + ".prg.fa").exists()
+        or Path(output_prefix + ".update_DS.zip").exists()
+        or Path(output_prefix + ".prg.bin").exists()
+        or Path(output_prefix + ".prg.bin.zip").exists()
+        or Path(output_prefix + ".prg.gfa").exists()
+        or Path(output_prefix + ".prg.gfa.zip").exists()
+    )
+
+
+def create_temp_dir(output_dir: Path) -> Path:
+    temp_dir = tempfile.mkdtemp(dir=output_dir)
+    return Path(temp_dir)
+
+
+def get_temp_dir_for_multiprocess(root_temp_dir: Path):
+    temp_dir = root_temp_dir / "mp_temp"
+    os.makedirs(temp_dir, exist_ok=True)
+    return temp_dir
+
+
+def zip_set_of_files(zip_filepath: Path, filename_to_filepath: Dict[str, Path]):
+    is_a_zip_file = zip_filepath.suffix == ".zip"
+    assert is_a_zip_file, "zip_set_of_files() was not given a .zip filepath"
+    with ZipFile(zip_filepath, "w") as zip_file:
+        for filename, filepath in filename_to_filepath.items():
+            zip_file.write(filepath, filename)
+
+
+# Note: this whole class is not unit tested
+class SetOutputFiles:
+    """
+    Class that represents all files that were generated in processing a PRG (PRG fasta, GFA, pickles, etc)
+    """
+    def __init__(self, PRG: Optional[Path] = None, binary_PRG: Optional[Path] = None,
+                 gfa: Optional[Path] = None, pickle: Optional[Path] = None,
+                 stats: Optional[Path] = None):
+        self.PRG: Optional[Path] = PRG
+        self.binary_PRG: Optional[Path] = binary_PRG
+        self.gfa: Optional[Path] = gfa
+        self.pickle: Optional[Path] = pickle
+        self.stats: Optional[Path] = stats
```

Comment:
> I've seen you do this before, specifying the type of the member attribute, but is that really necessary since the type of the function param is already specified?

Yeah, this is just for more documentation purposes, like when I `ctrl` click on the attribute in `pycharm`, and get taken to this line, and immediately see the type of the attribute. Not much necessary though as, as you said, it is also in the function param... Just more documentation!

> Also, because the __init__ is just setting attributes, you could use a dataclass here

Indeed, I was not aware of `dataclass` at all! Will switch to it!

Reply:
Switched to `dataclass` in https://github.com/iqbal-lab-org/make_prg/pull/34/commits/963185f294d4daf4d64a5f64abf7ec1c2f827618 . Also managed to switch class `ClusteringResult` to dataclass, as well as 3 other classes in the update code (not shown here). Just loved this std annotation, will make use of it a lot in the future, thanks!